gitlab/.gitlab-ci-main.yml

```diff
@@ -19,46 +23,80 @@ services:
   - "docker:dind"
 stages:
   - static
+  - preflight
   - integration
-  - audit
   - deploy
 variables:
   extends: .variables
   DOCKER_HOST: "tcp://localhost:2375"
+# @todo see GOVCMS-3110
+#cache:
+#  key: "$CI_PROJECT_PATH_SLUG"
+#  paths:
+#    - /home/.composer/cache
+
+# Static.
 
 vet:
-  <<: *static_template
   script:
-    - echo "We'll run verification here"
-static:
+    - composer validate
+  when: on_success
+  allow_failure: false
+  stage: static
+
+lint:
   <<: *static_template
-  extends: .static-tests
+  extends: .job-lint
+  # Run all lines together so that they all run even if any fail.
   script: |
-    set +euo pipefail # Make sure all scripts run (an error in any is still a gitlab job fail).
-    ./vendor/bin/govcms-lint web/modules/custom
-    ./vendor/bin/govcms-lint web/themes/custom
-    # @todo this needs to run only static phpunit tests - no integration tests as Drupal can't be bootstrapped.
-    # ./vendor/bin/govcms-phpunit --testsuite=unit
-  stage: static
-functional:
+    set +exuo pipefail # Linting...
+    docker-compose exec -T test bash -c '/app/vendor/bin/govcms-lint web/modules/custom'
+    docker-compose exec -T test bash -c '/app/vendor/bin/govcms-lint web/themes/custom'
+unit:
+  <<: *static_template
+  extends: .job-unit
+  script:
+    - docker-compose exec -T test bash -c './vendor/bin/govcms-unit --testsuite=unit'
+
+# Pre-flight.
+
+preflight:
+  <<: *bootstrap_template
+  extends: .job-preflight
+  stage: preflight
+  script:
+    - docker-compose exec -T cli bash -c 'drush cim -y'
+    - docker-compose exec -T cli bash -c 'drush st'
+    - docker-compose exec -T cli bash -c 'drush updatedb:status'
```

Comment:
I'm intentionally not shortcutting these commands. Keeping native and explicit so there is no confusion what they do. ..